_posts/2020-12-01-kafka-cloud-events.adoc

```diff
@@ -0,0 +1,185 @@
+
+The `data` attribute contains the wrapped business event.
+It is an essential part, and the other attributes are just providing details about that specific business event.
+
+You can also define _extensions_.
+These extensions would be a set of custom attributes used when the proposed set of attributes is not enough for your use case.
+
+That's it!
+So, we can summarize Cloud Events as _just enough metadata to understand an event_ - its source, an id, a type, and the business data.
+
+== Cloud Events on the wire - the bindings
+
+But, how would these events be encoded?
+The previous example using JSON is nice, but some protocols may want to leverage their own capabilities to transmit these metadata.
+
+That's why Cloud Events also proposes bindings.
+A binding is a set of recommendations specific to one protocol.
+It explains how each protocol should encode Cloud Events.
+For example, there is a binding for HTTP, one for Kafka, and another for AMQP.
+
+Most of these bindings propose two approaches:
+
+* structured
+* binary
+
+The structured approach keeps event metadata and data together in the payload of the message or request.
+It generally uses JSON to encode that data.
+If you pass the Cloud Event example (from above), in an HTTP request, it will use the structured mode.
+It will also use the structured mode when you write that JSON snippet in a Kafka record's value.
+
+The structured approach allows simple forwarding across multiple protocols.
+However, it may not be efficient and may constraint the type of business data.
+
+The other approach relies on protocol capabilities and enables efficient transfer and encoding.
+If we use the binary mode with Kafka, we will store the `data` attribute value in the Kafka record's value and pass the other attributes using the record's headers.
+Consequently, business data can be encoded using binary protocols such as Avro, leading to higher efficiency.
+
+The rest of the post explains how you can send and receive Cloud Events using Quarkus, Kafka, and Reactive Messaging.
+
+== Sending Cloud Events on Kafka
+
+The Kafka connector used by Quarkus has built-in support for Cloud Events.
+It can send and consume Cloud Events using the structured mode (encoding everything in a JSON payload) or the binary mode (using Kafka headers).
+
+To write your outgoing messages as Cloud Event, you only need to specify the `cloud-events-type` and `cloud-events-source` attributes on your channel:
+
+[source, properties]
+----
+mp.messaging.outgoing.generated-price.connector=smallrye-kafka
+mp.messaging.outgoing.generated-price.topic=prices
+mp.messaging.outgoing.generated-price.value.serializer=org.apache.kafka.common.serialization.IntegerSerializer
+mp.messaging.outgoing.generated-price.cloud-events-source=price-generators
+mp.messaging.outgoing.generated-price.cloud-events-type=price
+mp.messaging.outgoing.generated-price.cloud-events-subject=generated-prices
```

Comment:
Do you need to describe what the `subject` key is for?

Reply:
Oh, good catch! Forgot I added this one in my example.